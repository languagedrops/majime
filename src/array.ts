import { getRandom } from './random'


declare global {
  interface Array<T> {
    remove(elem: T): T[]
    flatten(): T
    removeLastElement(): T[]
    unique(): T[]
    toSet(): Set<T>
    reversed(): T[]
    insert(elem: T, index: number): T[]
    isUnique(): boolean
    substract(otherArray: T[]): T[]
    groupBy(keyExtractor: (item: T, index: number) => string): {[key: string]: T }
    groupByMultipleKeys(multipleKeysExtractor: (item: T, index: number) => string[]): {[key: string]: T }
    groupByAndMap<U>(keyExtractor: (item: T, index: number) => string, transform: (item: T, index: number) => U): {[key: string]: U }
    groupByMultipleValues(keyExtractor: (item: T, index: number) => string): {[key: string]: T[] }
    shuffle(): T[]
    randomElement(): T
    randomElementWithExceptions(except: T[]): T
    randomElements(count: number): T[]
    firstElement(): T
    lastElement(): T
    lastElements(numberOfElements: number): T[]
    filterNull(): Array<Exclude<T, null | undefined>>
  }
}


if (!Array.prototype.remove) {
  Array.prototype.remove = function<T>(elem: T): T[] {
    return this.filter((e) => e !== elem)
  }
}

if (!Array.prototype.flatten) {
  Array.prototype.flatten = function<T>(): T {
    return [].concat(...this) as any
  }
}

if (!Array.prototype.removeLastElement) {
  Array.prototype.removeLastElement = function<T>(): T[] {
    return removeLastElement(this)
  }
}

export const removeLastElement = <T>(fromArray: T[]): T[] => {
  if (fromArray.length === 0) {
    return fromArray
  }
  return fromArray.slice(0, fromArray.length - 1)
}

if (!Array.prototype.unique) {
  Array.prototype.unique = function<T>(): T[] {
    return [...new Set(this)]
  }
}

if (!Array.prototype.toSet) {
  Array.prototype.toSet = function<T>(): Set<T> {
    return new Set(this)
  }
}

if (!Array.prototype.reversed) {
  Array.prototype.reversed = function<T>(): T[] {
    return reverseArray(this)
  }
}

export const reverseArray = <T>(array: T[]): T[] => {
  return [...array].reverse()
}

if (!Array.prototype.insert) {
  Array.prototype.insert = function<T>(elem: T, index: number): T[] {
    return insert(this, index, elem)
  }
}


export const insert = <T>(array: T[], index: number, newItem: T): T[] => [
  ...array.slice(0, index),
  newItem,
  ...array.slice(index),
]

if (!Array.prototype.isUnique) {
  Array.prototype.isUnique = function(): boolean {
    return isArrayUnique(this)
  }
}

export const isArrayUnique = <T>(array: T[]): boolean => array.length === new Set(array).size


if (!Array.prototype.substract) {
  Array.prototype.substract = function<T>(otherArray: T[]): T[] {
    return substractArrays(this, otherArray)
  }
}

export const substractArrays = <T>(a: T[], b: T[]): T[] => {
  return a.filter((aElement) => !b.includes(aElement))
}

if (!Array.prototype.groupBy) {
  Array.prototype.groupBy = function<T>(keyExtractor: (item: T, index: number) => string): {[key: string]: T } {
    return groupBy(this, keyExtractor)
  }
}


export const groupBy = <T>(fromArray: T[], keyExtractor: (item: T, index: number) => string): {[key: string]: T } => {
  const groupped: {[key: string]: T } = {}
  fromArray.forEach( (item, index) => {
    groupped[keyExtractor(item, index)] = item
  })
  return groupped
}

if (!Array.prototype.groupByMultipleKeys) {
  Array.prototype.groupByMultipleKeys = function<T>(multipleKeysExtractor: (item: T, index: number) => string[]): {[key: string]: T } {
    return groupByMultipleKeys(this, multipleKeysExtractor)
  }
}

export const groupByMultipleKeys = <T>(fromArray: T[], multipleKeysExtractor: (item: T, index: number) => string[]): {[key: string]: T } => {
  const groupped: {[key: string]: T } = {}
  fromArray.forEach( (item, index) => {
    const keys = multipleKeysExtractor(item, index)
    keys.forEach((currentKey) => {
      groupped[currentKey] = item
    })
  })
  return groupped
}

if (!Array.prototype.groupByAndMap) {
  Array.prototype.groupByAndMap = function<T, U>(keyExtractor: (item: T, index: number) => string, transform: (item: T, index: number) => U): {[key: string]: U } {
    return groupByAndMap(this, keyExtractor, transform)
  }
}

export const groupByAndMap = <T, U>(fromArray: T[], keyExtractor: (item: T, index: number) => string, transform: (item: T, index: number) => U): {[key: string]: U } => {
  const groupped: {[key: string]: U } = {}
  fromArray.forEach( (item, index) => {
    groupped[keyExtractor(item, index)] = transform(item, index)
  })
  return groupped
}

if (!Array.prototype.groupByMultipleValues) {
  Array.prototype.groupByMultipleValues = function<T>(keyExtractor: (item: T, index: number) => string): {[key: string]: T[] } {
    return groupByMultipleValues(this, keyExtractor)
  }
}


export const groupByMultipleValues = <T>(fromArray: T[], keyExtractor: (item: T, index: number) => string): {[key: string]: T[] } => {
  const groupped: {[key: string]: T[] } = {}
  fromArray.forEach( (item, index) => {
    const key = keyExtractor(item, index)
    groupped[key] = [...(groupped[key] || []), item]
  })
  return groupped
}

if (!Array.prototype.shuffle) {
  Array.prototype.shuffle = function<T>(): T[] {
    return shuffle(this)
  }
}

export const shuffle = <T>(array: T[]): T[] => {
  // if it's 1 or 0 items, just return
  if (array.length <= 1) {
    return array
  }

  const newArray = array.slice()

  // For each index in array
  for (let i = 0; i < newArray.length; i++) {

    // choose a random not-yet-placed item to place there
    // must be an item AFTER the current item, because the stuff
    // before has all already been placed
    const randomChoiceIndex = getRandom(i, newArray.length - 1)

    const temp = newArray[i]
    newArray[i] = newArray[randomChoiceIndex]
    newArray[randomChoiceIndex] = temp
  }

  return newArray
}

if (!Array.prototype.randomElement) {
  Array.prototype.randomElement = function<T>(): T {
    return randomElement(this)
  }
}

export const randomElement = <T>(fromArray: T[]): T => {
  return fromArray[Math.floor(Math.random() * fromArray.length)]
}


if (!Array.prototype.randomElementWithExceptions) {
  Array.prototype.randomElementWithExceptions = function<T>(except: T[]): T {
    return randomElementWithExceptions(this, except)
  }
}

export const randomElementWithExceptions = <T>(fromArray: T[], excludeArray: T[]) => {
  const filteredArray = substractArrays(fromArray, excludeArray)
  return randomElement(filteredArray)
}


if (!Array.prototype.randomElements) {
  Array.prototype.randomElements = function<T>(count: number): T[] {
    return randomElements(this, count)
  }
}

export const randomElements = <T>(fromArray: T[], count: number): T[] => {
  if (count > fromArray.length) { throw Error('trying to get more elements than array length') }
  return shuffle(fromArray).slice(0, count)
}


if (!Array.prototype.firstElement) {
  Array.prototype.firstElement = function<T>(): T {
    return firstElement(this)
  }
}

export const firstElement = <T>(fromArray: T[]): T => {
  return fromArray[0]
}


if (!Array.prototype.lastElement) {
  Array.prototype.lastElement = function<T>(): T {
    return lastElement(this)
  }
}

export const lastElement = <T>(array: T[]): T => {
  return array[array.length - 1]
}


if (!Array.prototype.lastElements) {
  Array.prototype.lastElements = function<T>(numberOfElements: number): T[] {
    return lastElements(this, numberOfElements)
  }
}

export const lastElements = <T>(array: T[], numberOfElements: number): T[] => {
  return array.slice(Math.max(array.length - numberOfElements, 0))
}

if (!Array.prototype.filterNull) {
  Array.prototype.filterNull = function<T>(): Array<Exclude<T, null | undefined>> {
    return filterNull(this)
  }
}


export const filterNull = <T>(array: T[]): Array<Exclude<T, null | undefined>> => {
  return array.filter( (elem) => elem != null) as Array<Exclude<T, null | undefined>>
}


