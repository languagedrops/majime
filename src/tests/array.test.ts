import { sortedByProperty, chunk, sortIntersection } from '../array'

describe('Array function', () => {

  describe('removeLastElement', () => {
    it('Should correctly delete last element', () => {
      expect([1, 2, 3, 4].removeLastElement()).toEqual([1, 2, 3])
      expect([1].removeLastElement()).toEqual([])
      expect([].removeLastElement()).toEqual([])
    })
  })

  describe('sortArrayByFunction', () => {
    it('Should sort the arrays by the length of the name while keeping the equal ones in the same order', () => {
      const sourceArray = [
        { name: 'J. K. Rowling', id: 3 },
        { name: 'Agatha Christie', id: 4 },
        { name: 'Jane Austen', id: 1 },
        { name: 'George Orwell', id: 2 },
      ]
      const orderedArray = [
        { name: 'Jane Austen', id: 1 },
        { name: 'J. K. Rowling', id: 3 },
        { name: 'George Orwell', id: 2 },
        { name: 'Agatha Christie', id: 4 },
      ]
      const compareFunction = (element: any) => element.name.length
      expect(sortedByProperty(sourceArray, compareFunction)).toEqual(orderedArray)
    })

    it('Should sort the arrays by the length of the name in reverse order while keeping the equal ones in the same order', () => {
      const sourceArray = [
        { name: 'J. K. Rowling', id: 3 },
        { name: 'Agatha Christie', id: 4 },
        { name: 'Jane Austen', id: 1 },
        { name: 'George Orwell', id: 2 },
      ]
      const orderedArray = [
        { name: 'Agatha Christie', id: 4 },
        { name: 'J. K. Rowling', id: 3 },
        { name: 'George Orwell', id: 2 },
        { name: 'Jane Austen', id: 1 },
      ]
      const compareFunction = (element: any) => element.name.length
      expect(sortedByProperty(sourceArray, compareFunction, true)).toEqual(orderedArray)
    })
  })

  describe('Chunk', () => {
    it('Should split array into chunks', () => {
      const testArray = [1, 2, 3, 4, 5, 6, 7]
      const expectedArray = [[1, 2], [3, 4], [5, 6], [7]]
      expect(chunk(2, testArray)).toEqual(expectedArray)
    })
  })

  describe('reorderArrayByPrevious', () => {
    it('Should return source like aray if order array didn\'t provide', () => {
      const sourceArray = ['a', 'b', 'c', 'd', 'e']
      expect(sortIntersection(sourceArray)).toEqual(sourceArray)
    })

    it('Should return source like array if length of array isn\'t ste same', () => {
      const sourceArray = ['a', 'b', 'c', 'd', 'e']
      const orderArray = ['a', 'b', 'f']
      expect(sortIntersection(sourceArray, orderArray)).toEqual(sourceArray)
    })

    it('should return source like array if there aren\'t common components in both arrays', () => {
      const sourceArray = ['a', 'b', 'c', 'd', 'e']
      const orderArray = ['f', 'n', 'm', 'h', 'j']
      expect(sortIntersection(sourceArray, orderArray)).toEqual(sourceArray)
    })

    it('Should return array where common components will be at the same place as they are in order array', () => {
      const sourceArray = ['a', 'b', 'c', 'd', 'e']
      const orderArray = ['f', 'c', 'm', 'e', 'b']
      const result = sortIntersection(sourceArray, orderArray)
      expect(result[1]).toBe('c')
      expect(result[3]).toBe('e')
      expect(result[4]).toBe('b')
    })

    it('Should return order like array, if all elements are the same', () => {
      const sourceArray = ['a', 'b', 'c', 'd', 'e']
      const orderArray = ['b', 'e', 'd', 'a', 'c']
      expect(sortIntersection(sourceArray, orderArray)).toEqual(orderArray)
    })
  })

  describe('removeLastElement', () => {
    it('Should correctly delete last element', () => {
      expect([1, 2, 3, 4].removeLastElement()).toEqual([1, 2, 3])
      expect([1].removeLastElement()).toEqual([])
      expect([].removeLastElement()).toEqual([])
    })
  })

  describe('takeWhile', () =>Â {
    it('Should correctly return the elements while they are matching the condition', () => {
      expect([1, 2, 3, 2, 3, 4, 5, 2].takeWhile((element) => element < 3)).toEqual([1, 2])
      expect([1, 2].takeWhile((element) => element < 3)).toEqual([1, 2])
      expect([1].takeWhile((element) => element < 3)).toEqual([1])
      expect([].takeWhile((element) => element < 3)).toEqual([])
      expect([1, 2].takeWhile((element) => element > 3)).toEqual([])
    })

    it('Should correctly return the elements while they are matching the condition with reverse', () => {
      expect([1, 2, 3, 2, 3, 4, 5, 2].takeWhile((element) => element < 3, true)).toEqual([2])
      expect([1, 2].takeWhile((element) => element < 3, true)).toEqual([2, 1])
      expect([1].takeWhile((element) => element < 3, true)).toEqual([1])
      expect([].takeWhile((element) => element < 3, true)).toEqual([])
      expect([1, 2].takeWhile((element) => element > 3, true)).toEqual([])
    })

  })
})
