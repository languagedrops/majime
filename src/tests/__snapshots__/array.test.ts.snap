// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array function Unique Shouldn't change order of array 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`Array function Unique Shouldn't change order of array 2`] = `
Array [
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`Array function Unique Shouldn't change order of array 3`] = `
Array [
  "test1",
  "test2",
  "test3",
  "test4",
  "test5",
]
`;

exports[`Array function Unique Shouldn't change order of array 4`] = `
Array [
  "test1",
  "test2",
  "test3",
  "test4",
  "test5",
]
`;
